kube/src/client/mod.rs

```diff
@@ -43,6 +43,7 @@ use std::convert::{TryFrom, TryInto};
 #[derive(Clone)]
 pub struct Client {
     cluster_url: url::Url,
+    // REVIEW `default_ns` is never used?
```

Comment:
The `Config` exposes it so people can create `Api::namespaced` with the namespace from the configured `default_ns`. It is awkward though.

Reply:
Ah, I see. So something like this?
```rust
let config = Config::infer()?;
let client = Client::new(config.clone());
let pods: Api<Pod> = Api::namespaced(client, &config.default_ns);
```

Any reasons to store it in `Client`? It's private, and I don't think it's used.

Reply:
Well, it's just there so the Config can absorb it. And it's currently in the Client because that's where things that read the incluster config is :/

Reply:
Actually, I am wrong. I thought it needed there for something, but it's possible I only put it there because I thought I could make a better shorthand for `Api::namespaced` if the default_ns was set, but that never happened. We can remove it.

Reply:
I found [the commit](https://github.com/clux/kube-rs/commit/cde680d2be1280bd9733b136ee53535df04f2f90) you added it. So it's for stashed `Api::default_namespaced(client: Client)` (https://github.com/clux/kube-rs/issues/209#issuecomment-607067454).

---

I started skimming `tower` docs (they just released [v0.4](https://docs.rs/tower/0.4.1/tower/index.html)) and thinking of ways to break things up nicely.

[`Layer`](https://docs.rs/tower/0.4.1/tower/trait.Layer.html) (middleware) should allow moving a lot out of the core (TLS, timeout, proxy, compression, auth, logging, tracing, etc.) and allow users to just include what they want. Though, I'm not sure if it's possible to have a single `tower::Service` that can handle both HTTP request/response and WebSocket connection (different response types). I haven't understood how it works yet, so I'm hoping it's possible.

Reply:
Ah, I even had an issue :+1:

> Tower

Yeah, I am not very knowledgeable on this front either, just started reading a bit of the 0.4 docs. Gut feel is that it's going to be hard to make the model fit when we have: regular http req/res, streaming http with long polling (and built in restarts with custom behaviour), persistent ws. Not sure if it's actually the right way to go for us.

Reply:
But then it could potentially make the whole client passing a lot less awkward and give us more testable interfaces. Definitely have a go. It would also be interesting to see how it would interact with the whole oauth2 situation.

Reply:
> streaming http with long polling (and built in restarts with custom behaviour)

Can you tell me which method this is?

Now with WebSocket unified, it should be possible to do all of them from `Response<B: http_body::Body>`.

- text: concatenate bytes from stream
- typed: text + deserialize
- streaming: the response body is streaming
- WebSocket stream: wrap in WebSocket Stream after upgrading